airflow/jobs/job.py

```diff
@@ -117,6 +115,15 @@ def __init__(self, executor=None, heartrate=None, **kwargs):
     def executor(self):
         return ExecutorLoader.get_default_executor()
 
+    @cached_property
+    def heartrate(self):
+        if self.job_type == "TriggererJob":
+            return conf.getfloat("triggerer", "JOB_HEARTBEAT_SEC")
+        else:
+            # Heartrate used to be hardcoded to scheduler, so in all other
+            # cases continue to use that value for back compat
+            return conf.getfloat("scheduler", "JOB_HEARTBEAT_SEC")
```

Comment:
This breaks line 110/108 unfortunately, where `self.heartrate` is assigned to. It seems to me this switching logic can actually be move the the runner though.

Reply:
Good catch! Interesting that no tests failed for that. I'll check `self` for a heartbeat first before fetching from the config

Reply:
Now that I think about it, I think likely no tests failed because we never pass a value to the constructor that isn't itself from config. Since after reading it back from the DB it can _only_ be the config value.

I'll still add a test and code to cover it though.

Like I mentioned in the other comment thread, I'm trying to scope this down to a bug fix. I'll cut another Issue to investigate refactoring heartrate entirely (I think adding a column for it in Job on the DB would be best).

Reply:
After looking deeper, I don't think this breaks those lines at all. If a value is assigned to heartbeat directly in the constructor then the cached_property is not used. `self.executor` is already working in this way if you look above.

I added a unit test to confirm this and also checked manually with a dummy class in a python session:

![Screenshot from 2023-08-04 11-26-38](https://github.com/apache/airflow/assets/65743084/f890a7a6-cf5b-4def-a717-4e4c1cb8dd85)

Reply:
I tested it yesterday:
When we initiate the Job from the CLI, we provide the heartbeat and it will be used. But when we load the row from the Metadata, the heartbeat will be None, in this case we will use the cached property.
I don't see any issue with this approach.

Reply:
Actually this is ok accordig to how cached_property works. I recently stumbled upon that.

Cached_property works different than @property -> it actually creates an attribute that is set the first time the property is used and you can freely delete or assign the attribute a different value. This is a nice one, because you can even invalidate the cache by simply deleting the attribute (which is for example very cool for testing and I am already using it to invalidat sensitive field list when loading provider configuration https://github.com/apache/airflow/blob/main/airflow/configuration.py#L1854

It was working like that since 3.8 but it made it into "official part of the API" in Python 3.9. Similarly as HashMap ordering graduated from being implementation detail to official behaviour.

Reply:
> The mechanics of [cached_property()](https://docs.python.org/3/library/functools.html#functools.cached_property) are somewhat different from [property()](https://docs.python.org/3/library/functions.html#property). A regular property blocks attribute writes unless a setter is defined. In contrast, a cached_property allows writes.

https://docs.python.org/3/library/functools.html#functools.cached_property

So I guess this is good then.